// md-editor-v3/lib/es/chunks/index3.mjs
import { useId as Fe, provide as I, computed as A, reactive as re, watch as E, onMounted as H, ref as G, inject as C, nextTick as U, shallowRef as ee, onBeforeUnmount as te, toRef as Ae, defineComponent as ce, createVNode as j, Fragment as Re, h as He } from "vue";
import { deepMerge as Me, deepClone as Pe, randomId as le, debounce as Le } from "@vavt/util";
import { p as f, g as P, f as ue, s as me, b as Oe, a as _e } from "./config.mjs";
import { b as v, U as Be, C as Ee, B as X, O as ke, E as Q, P as fe, F as he, f as pe, g as ge, H as ve, h as ye, R as ie, e as je, i as de, d as Ne, j as oe, k as De, T as Ve } from "./event-bus.mjs";
import { a as O, u as Ge, S as Ue, c as ze, z as We } from "./dom.mjs";
import qe from "@vavt/copy2clipboard";
import Ke from "markdown-it";
import Ze from "markdown-it-image-figures";
import Je from "markdown-it-sub";
import Xe from "markdown-it-sup";
import { g as Ye } from "./index5.mjs";
import Qe from "medium-zoom";
const M = {
  hljs: `${f}-hljs`,
  hlcss: `${f}-hlCss`,
  prettier: `${f}-prettier`,
  prettierMD: `${f}-prettierMD`,
  cropperjs: `${f}-cropper`,
  croppercss: `${f}-cropperCss`,
  screenfull: `${f}-screenfull`,
  mermaidM: `${f}-mermaid-m`,
  mermaid: `${f}-mermaid`,
  katexjs: `${f}-katex`,
  katexcss: `${f}-katexCss`,
  echarts: `${f}-echarts`
}, Zt = (e, t, r) => {
  const { editorId: n } = r, o = re({
    // 是否已编译成html
    buildFinished: !1,
    // 存储当前最新的html
    html: ""
  });
  E(
    () => e.modelValue,
    () => {
      o.buildFinished = !1;
    }
  ), H(() => {
    v.on(n, {
      name: X,
      callback(m) {
        o.buildFinished = !0, o.html = m;
      }
    }), v.on(n, {
      name: ke,
      callback() {
        const m = new Promise((d) => {
          if (o.buildFinished)
            d(o.html);
          else {
            const u = (s) => {
              d(s), v.remove(n, X, u);
            };
            v.on(n, {
              name: X,
              callback: u
            });
          }
        });
        e.onSave ? e.onSave(e.modelValue, m) : t.emit("onSave", e.modelValue, m);
      }
    });
  });
}, Te = (e, {
  editorId: t,
  rootRef: r,
  setting: n
}) => {
  const o = P.editorExtensions.highlight, m = P.editorExtensionsAttrs.highlight;
  I("editorId", t), I("rootRef", r), I(
    "theme",
    A(() => e.theme)
  ), I(
    "language",
    A(() => e.language)
  ), I(
    "highlight",
    A(() => {
      const { js: u } = o, s = {
        ...ue,
        ...o.css
      }, { js: l, css: i = {} } = m || {}, a = e.codeStyleReverse && e.codeStyleReverseList.includes(e.previewTheme) ? "dark" : e.theme, c = s[e.codeTheme] ? s[e.codeTheme][a] : ue.atom[a], $ = s[e.codeTheme] && i[e.codeTheme] ? i[e.codeTheme][a] : i.atom ? i.atom[a] : {};
      return {
        js: {
          src: u,
          ...l
        },
        css: {
          href: c,
          ...$
        }
      };
    })
  ), I("showCodeRowNumber", e.showCodeRowNumber);
  const d = A(() => {
    const u = {
      ...me,
      ...P.editorConfig.languageUserDefined
    };
    return Me(
      Pe(me["en-US"]),
      u[e.language] || {}
    );
  });
  return I("usedLanguageText", d), I(
    "previewTheme",
    A(() => e.previewTheme)
  ), I(
    "customIcon",
    A(() => e.customIcon)
  ), I(
    "setting",
    A(() => n ? {
      // setting是reactive，不转化是可以直接赋值的
      ...n
    } : {
      preview: !0,
      htmlPreview: !1,
      previewOnly: !1,
      pageFullscreen: !1,
      fullscreen: !1
    })
  ), { editorId: t };
}, Jt = (e, t) => (I("tabWidth", e.tabWidth), I(
  "disabled",
  A(() => e.disabled)
), I(
  "showToolbarName",
  A(() => e.showToolbarName)
), I("noUploadImg", e.noUploadImg), I(
  "tableShape",
  A(() => e.tableShape)
), I("noPrettier", e.noPrettier), I(
  "codeTheme",
  A(() => e.codeTheme)
), I("updateSetting", t.updateSetting), I(
  "catalogVisible",
  A(() => t.catalogVisible.value)
), I("defToolbars", t.defToolbars), I(
  "floatingToolbars",
  A(() => e.floatingToolbars)
), Te(e, t)), Xt = (e) => {
  const { noPrettier: t, noUploadImg: r } = e, { editorExtensions: n, editorExtensionsAttrs: o } = P, m = t || n.prettier.prettierInstance, d = t || n.prettier.parserMarkdownInstance, u = r || n.cropper.instance;
  H(() => {
    if (!u) {
      const { js: s = {}, css: l = {} } = o.cropper || {};
      O("link", {
        ...l,
        rel: "stylesheet",
        href: n.cropper.css,
        id: M.croppercss
      }), O("script", {
        ...s,
        src: n.cropper.js,
        id: M.cropperjs
      });
    }
    if (!m) {
      const { standaloneJs: s = {} } = o.prettier || {};
      O("script", {
        ...s,
        src: n.prettier.standaloneJs,
        id: M.prettier
      });
    }
    if (!d) {
      const { parserMarkdownJs: s = {} } = o.prettier || {};
      O("script", {
        ...s,
        src: n.prettier.parserMarkdownJs,
        id: M.prettierMD
      });
    }
  });
}, Yt = (e, t, r) => {
  const { editorId: n } = r;
  H(() => {
    v.on(n, {
      name: Q,
      callback: (o) => {
        e.onError?.(o), t.emit("onError", o);
      }
    });
  });
}, Qt = (e, t, r) => {
  const { editorId: n } = r, o = re({
    pageFullscreen: e.pageFullscreen,
    fullscreen: !1,
    preview: e.preview,
    htmlPreview: e.preview ? !1 : e.htmlPreview,
    previewOnly: !1
  }), m = re({ ...o }), d = (l, i) => {
    const a = i === void 0 ? !o[l] : i;
    switch (l) {
      case "preview": {
        o.htmlPreview = !1, o.previewOnly = !1;
        break;
      }
      case "htmlPreview": {
        o.preview = !1, o.previewOnly = !1;
        break;
      }
      case "previewOnly": {
        a ? !o.preview && !o.htmlPreview && (o.preview = !0) : (m.preview || (o.preview = !1), m.htmlPreview || (o.htmlPreview = !1));
        break;
      }
    }
    m[l] = a, o[l] = a;
  };
  let u = "";
  const s = () => {
    o.pageFullscreen || o.fullscreen ? document.body.style.overflow = "hidden" : document.body.style.overflow = u;
  };
  return E(() => [o.pageFullscreen, o.fullscreen], s), H(() => {
    v.on(n, {
      name: Be,
      callback(l, i) {
        const a = (c) => {
          v.emit(n, ie, "image", {
            desc: "",
            urls: c
          }), i?.();
        };
        e.onUploadImg ? e.onUploadImg(l, a) : t.emit("onUploadImg", l, a);
      }
    }), u = document.body.style.overflow, s();
  }), [o, d];
}, en = (e, t) => {
  const { editorId: r } = t, n = G(!1);
  return H(() => {
    v.on(r, {
      name: Ee,
      callback: (o) => {
        o === void 0 ? n.value = !n.value : n.value = o;
      }
    });
  }), n;
}, tn = (e, t, r) => {
  const { editorId: n, catalogVisible: o, setting: m, updateSetting: d, codeRef: u } = r;
  E(
    () => m.pageFullscreen,
    (l) => {
      v.emit(n, fe, l);
    }
  ), E(
    () => m.fullscreen,
    (l) => {
      v.emit(n, he, l);
    }
  ), E(
    () => m.preview,
    (l) => {
      v.emit(n, pe, l);
    }
  ), E(
    () => m.previewOnly,
    (l) => {
      v.emit(n, ge, l);
    }
  ), E(
    () => m.htmlPreview,
    (l) => {
      v.emit(n, ve, l);
    }
  ), E(o, (l) => {
    v.emit(n, ye, l);
  });
  const s = {
    on(l, i) {
      switch (l) {
        case "pageFullscreen": {
          v.on(n, {
            name: fe,
            callback(a) {
              i(a);
            }
          });
          break;
        }
        case "fullscreen": {
          v.on(n, {
            name: he,
            callback(a) {
              i(a);
            }
          });
          break;
        }
        case "preview": {
          v.on(n, {
            name: pe,
            callback(a) {
              i(a);
            }
          });
          break;
        }
        case "previewOnly": {
          v.on(n, {
            name: ge,
            callback(a) {
              i(a);
            }
          });
          break;
        }
        case "htmlPreview": {
          v.on(n, {
            name: ve,
            callback(a) {
              i(a);
            }
          });
          break;
        }
        case "catalog": {
          v.on(n, {
            name: ye,
            callback(a) {
              i(a);
            }
          });
          break;
        }
      }
    },
    togglePageFullscreen(l) {
      d("pageFullscreen", l);
    },
    toggleFullscreen(l) {
      v.emit(n, Ne, l);
    },
    togglePreview(l) {
      d("preview", l);
    },
    togglePreviewOnly(l) {
      d("previewOnly", l);
    },
    toggleHtmlPreview(l) {
      d("htmlPreview", l);
    },
    toggleCatalog(l) {
      v.emit(n, Ee, l);
    },
    triggerSave() {
      v.emit(n, ke);
    },
    insert(l) {
      v.emit(n, ie, "universal", { generate: l });
    },
    focus(l) {
      u.value?.focus(l);
    },
    rerender() {
      v.emit(n, de);
    },
    getSelectedText() {
      return u.value?.getSelectedText();
    },
    resetHistory() {
      u.value?.resetHistory();
    },
    domEventHandlers(l) {
      v.emit(n, je, l);
    },
    execCommand(l) {
      v.emit(n, ie, l);
    },
    getEditorView() {
      return u.value?.getEditorView();
    }
  };
  t.expose(s);
}, et = (e) => {
  const t = Fe();
  return e.id || e.editorId || `${f}-${t}`;
}, tt = (e, t, r) => {
  const n = C("editorId"), o = C("rootRef"), m = C("usedLanguageText"), d = C("setting"), u = () => {
    o.value.querySelectorAll(`#${n} .${f}-preview .${f}-code`).forEach((i) => {
      let a = -1;
      const c = i.querySelector(
        `.${f}-copy-button:not([data-processed])`
      );
      c && (c.onclick = ($) => {
        $.preventDefault(), clearTimeout(a);
        const g = (i.querySelector("input:checked + pre code") || i.querySelector("pre code")).textContent, { text: h, successTips: T, failTips: y } = m.value.copyCode;
        let b = T;
        qe(e.formatCopiedText(g || "")).catch(() => {
          b = y;
        }).finally(() => {
          c.dataset.isIcon ? c.dataset.tips = b : c.innerHTML = b, a = window.setTimeout(() => {
            c.dataset.isIcon ? c.dataset.tips = h : c.innerHTML = h;
          }, 1500);
        });
      }, c.setAttribute("data-processed", "true"));
    });
  }, s = () => {
    U(u);
  }, l = (i) => {
    i && U(u);
  };
  E([t, r], s), E(() => d.value.preview, l), E(() => d.value.htmlPreview, l), H(u);
}, nt = (e) => {
  const t = C("editorId"), r = C("theme"), n = C("rootRef"), { editorExtensions: o, editorExtensionsAttrs: m, echartsConfig: d } = P;
  let u = o.echarts.instance;
  const s = ee(-1), l = () => {
    !e.noEcharts && u && (s.value = s.value + 1);
  };
  E(
    () => r.value,
    () => {
      l();
    }
  ), H(() => {
    if (e.noEcharts || u)
      return;
    const g = o.echarts.js;
    O(
      "script",
      {
        ...m.echarts?.js,
        src: g,
        id: M.echarts,
        onload() {
          u = window.echarts, l();
        }
      },
      "echarts"
    );
  });
  let i = [], a = [], c = [];
  const $ = (g = !1) => {
    if (!i.length) {
      g && (a.forEach((b) => {
        b.dispose?.();
      }), c.forEach((b) => {
        b.disconnect?.();
      }), a = [], c = []);
      return;
    }
    const h = [], T = [], y = [];
    i.forEach((b, S) => {
      const w = a[S], F = c[S];
      if (g || !b || !b.isConnected || (n?.value ? !n.value.contains(b) : !1)) {
        w?.dispose?.(), F?.disconnect?.();
        return;
      }
      h.push(b), w && T.push(w), F && y.push(F);
    }), i = h, a = T, c = y;
  }, p = () => {
    $(), !e.noEcharts && u && Array.from(
      n.value.querySelectorAll(
        `#${t} div.${f}-echarts:not([data-processed])`
      )
    ).forEach((h) => {
      if (h.dataset.closed === "false")
        return !1;
      try {
        const T = o.echarts.parseOption(h.innerText, {
          editorId: t,
          element: h
        }), y = d(T), b = u.init(h, r.value);
        b.setOption(y), h.setAttribute("data-processed", ""), i.push(h), a.push(b);
        const S = new ResizeObserver(() => {
          b.resize();
        });
        S.observe(h), c.push(S);
      } catch (T) {
        v.emit(t, Q, {
          name: "echarts",
          message: T?.message,
          error: T
        });
      }
    });
  };
  return te(() => {
    $(!0);
  }), { reRenderEcharts: s, replaceEcharts: p };
}, ot = (e) => {
  const t = C("highlight"), r = ee(P.editorExtensions.highlight.instance);
  return H(() => {
    e.noHighlight || r.value || (O("link", {
      ...t.value.css,
      rel: "stylesheet",
      id: M.hlcss
    }), O(
      "script",
      {
        ...t.value.js,
        id: M.hljs,
        onload() {
          r.value = window.hljs;
        }
      },
      "hljs"
    ));
  }), E(
    () => t.value.css,
    () => {
      e.noHighlight || P.editorExtensions.highlight.instance || Ge("link", {
        ...t.value.css,
        rel: "stylesheet",
        id: M.hlcss
      });
    }
  ), r;
}, at = (e) => {
  const t = ee(P.editorExtensions.katex.instance);
  return H(() => {
    if (e.noKatex || t.value)
      return;
    const { editorExtensions: r, editorExtensionsAttrs: n } = P;
    O(
      "script",
      {
        ...n.katex?.js,
        src: r.katex.js,
        id: M.katexjs,
        onload() {
          t.value = window.katex;
        }
      },
      "katex"
    ), O("link", {
      ...n.katex?.css,
      rel: "stylesheet",
      href: r.katex.css,
      id: M.katexcss
    });
  }), t;
};
class rt {
  constructor(t) {
    this.options = t;
  }
  cache = /* @__PURE__ */ new Map();
  isExpired(t) {
    return t.expiresAt <= Date.now();
  }
  deleteExpired(t, r) {
    const n = r ?? this.cache.get(t);
    return !n || !this.isExpired(n) ? !1 : (this.cache.delete(t), !0);
  }
  trim() {
    for (const [t, r] of this.cache)
      if (!this.deleteExpired(t, r))
        break;
    for (; this.cache.size > this.options.max; ) {
      const t = this.cache.keys().next().value;
      if (t === void 0)
        break;
      this.cache.delete(t);
    }
  }
  get(t) {
    const r = this.cache.get(t);
    if (!(!r || this.deleteExpired(t, r)))
      return this.cache.delete(t), this.cache.set(t, r), r.value;
  }
  set(t, r) {
    this.cache.delete(t), this.cache.set(t, {
      value: r,
      expiresAt: Date.now() + this.options.ttl
    }), this.trim();
  }
  clear() {
    this.cache.clear();
  }
}
const Y = new rt({
  max: 1e3,
  // 缓存10分钟
  ttl: 6e5
}), lt = (e) => {
  const t = C("editorId"), r = C("theme"), n = C("rootRef"), { editorExtensions: o, editorExtensionsAttrs: m, mermaidConfig: d } = P;
  let u = o.mermaid.instance;
  const s = ee(-1), l = () => {
    if (!e.noMermaid && u) {
      const a = r.value === "dark" ? {
        startOnLoad: !1,
        theme: "dark"
      } : {
        startOnLoad: !1,
        theme: "base",
        themeVariables: {
          background: "#ffffff",
          primaryColor: "#ffffff",
          primaryTextColor: "#1f2329",
          primaryBorderColor: "#b7c0cc",
          secondaryColor: "#f7f8fa",
          tertiaryColor: "#f7f8fa",
          lineColor: "#596273",
          edgeLabelBackground: "#ffffff",
          clusterBkg: "#ffffff",
          clusterBorder: "#b7c0cc"
        }
      };
      u.initialize(d(a)), s.value = s.value + 1;
    }
  };
  return E(
    () => r.value,
    () => {
      Y.clear(), l();
    }
  ), H(() => {
    if (e.noMermaid || u)
      return;
    const a = o.mermaid.js;
    /\.mjs/.test(a) ? (O("link", {
      ...m.mermaid?.js,
      rel: "modulepreload",
      href: a,
      id: M.mermaidM
    }), import(
      /* @vite-ignore */
      /* webpackIgnore: true */
      a
    ).then((c) => {
      u = c.default, l();
    }).catch((c) => {
      v.emit(t, Q, {
        name: "mermaid",
        message: `Failed to load mermaid module: ${c.message}`,
        error: c
      });
    })) : O(
      "script",
      {
        ...m.mermaid?.js,
        src: a,
        id: M.mermaid,
        onload() {
          u = window.mermaid, l();
        }
      },
      "mermaid"
    );
  }), { reRenderRef: s, replaceMermaid: async () => {
    if (!e.noMermaid && u) {
      const a = n.value.querySelectorAll(
        `div.${f}-mermaid`
      ), c = document.createElement("div"), $ = document.body.offsetWidth > 1366 ? document.body.offsetWidth : 1366, p = document.body.offsetHeight > 768 ? document.body.offsetHeight : 768;
      c.style.width = $ + "px", c.style.height = p + "px", c.style.position = "fixed", c.style.zIndex = "-10000", c.style.top = "-10000";
      let g = a.length;
      g > 0 && document.body.appendChild(c), await Promise.allSettled(
        Array.from(a).map((h) => (async (y) => {
          if (y.dataset.closed === "false")
            return !1;
          const b = y.innerText;
          let S = Y.get(b);
          if (!S) {
            const w = le();
            let F = { svg: "" };
            try {
              F = await u.render(w, b, c), S = await e.sanitizeMermaid(F.svg);
              const x = document.createElement("p");
              x.className = `${f}-mermaid`, x.setAttribute("data-processed", ""), x.setAttribute("data-content", b), x.innerHTML = S, x.children[0]?.removeAttribute("height"), Y.set(b, x.innerHTML), y.dataset.line !== void 0 && (x.dataset.line = y.dataset.line), y.replaceWith(x);
            } catch (x) {
              v.emit(t, Q, {
                name: "mermaid",
                message: x.message,
                error: x
              });
            }
            --g === 0 && c.remove();
          }
        })(h))
      );
    }
  } };
}, it = (e, t) => {
  t = t || {};
  const r = 3, n = t.marker || "!", o = n.charCodeAt(0), m = n.length;
  let d = "", u = "";
  const s = (i, a, c, $, p) => {
    const g = i[a];
    return g.type === "admonition_open" ? i[a].attrPush([
      "class",
      `${f}-admonition ${f}-admonition-${g.info}`
    ]) : g.type === "admonition_title_open" && i[a].attrPush(["class", `${f}-admonition-title`]), p.renderToken(i, a, c);
  }, l = (i) => {
    const a = i.trim().split(" ", 2);
    u = "", d = a[0], a.length > 1 && (u = i.substring(d.length + 2));
  };
  e.block.ruler.before(
    "code",
    "admonition",
    (i, a, c, $) => {
      let p, g, h, T = !1, y = i.bMarks[a] + i.tShift[a], b = i.eMarks[a];
      if (o !== i.src.charCodeAt(y))
        return !1;
      for (p = y + 1; p <= b && n[(p - y) % m] === i.src[p]; p++)
        ;
      const S = Math.floor((p - y) / m);
      if (S !== r)
        return !1;
      p -= (p - y) % m;
      const w = i.src.slice(y, p), F = i.src.slice(p, b);
      if (l(F), $)
        return !0;
      for (g = a; g++, !(g >= c || (y = i.bMarks[g] + i.tShift[g], b = i.eMarks[g], y < b && i.sCount[g] < i.blkIndent)); )
        if (o === i.src.charCodeAt(y) && !(i.sCount[g] - i.blkIndent >= 4)) {
          for (p = y + 1; p <= b && n[(p - y) % m] === i.src[p]; p++)
            ;
          if (!(Math.floor((p - y) / m) < S) && (p -= (p - y) % m, p = i.skipSpaces(p), !(p < b))) {
            T = !0;
            break;
          }
        }
      const x = i.parentType, L = i.lineMax;
      return i.parentType = "root", i.lineMax = g, h = i.push("admonition_open", "div", 1), h.markup = w, h.block = !0, h.info = d, h.map = [a, g], u && (h = i.push("admonition_title_open", "p", 1), h.markup = w + " " + d, h.map = [a, g], h = i.push("inline", "", 0), h.content = u, h.map = [a, i.line - 1], h.children = [], h = i.push("admonition_title_close", "p", -1), h.markup = w + " " + d), i.md.block.tokenize(i, a + 1, g), h = i.push("admonition_close", "div", -1), h.markup = i.src.slice(y, p), h.block = !0, i.parentType = x, i.lineMax = L, i.line = g + (T ? 1 : 0), !0;
    },
    {
      alt: ["paragraph", "reference", "blockquote", "list"]
    }
  ), e.renderer.rules.admonition_open = s, e.renderer.rules.admonition_title_open = s, e.renderer.rules.admonition_title_close = s, e.renderer.rules.admonition_close = s;
}, se = (e, t) => {
  const r = e.attrs ? e.attrs.slice() : [];
  return t.forEach((n) => {
    const o = e.attrIndex(n[0]);
    o < 0 ? r.push(n) : (r[o] = r[o].slice(), r[o][1] += ` ${n[1]}`);
  }), r;
}, st = (e, t) => {
  const r = e.renderer.rules.fence, n = e.utils.unescapeAll, o = /\[(\w*)(?::([\w ]*))?\]/, m = /::(open|close)/, d = (a) => a.info ? n(a.info).trim() : "", u = (a) => {
    const c = d(a), [$ = null, p = ""] = (o.exec(c) || []).slice(1);
    return [$, p];
  }, s = (a) => {
    const c = d(a);
    return c ? c.split(/(\s+)/g)[0] : "";
  }, l = (a) => {
    const c = a.info.match(m) || [], $ = c[1] === "open" || c[1] !== "close" && t.codeFoldable && a.content.trim().split(`
`).length < t.autoFoldThreshold, p = c[1] || t.codeFoldable ? "details" : "div", g = c[1] || t.codeFoldable ? "summary" : "div";
    return { open: $, tagContainer: p, tagHeader: g };
  }, i = (a, c, $, p, g) => {
    if (a[c].hidden)
      return "";
    const h = t.usedLanguageTextRef.value?.copyCode.text, T = t.customIconRef.value.copy || h, y = !!t.customIconRef.value.copy, b = `<span class="${f}-collapse-tips">${Ue("collapse-tips", t.customIconRef.value)}</span>`, [S] = u(a[c]);
    if (S === null) {
      const { open: k, tagContainer: R, tagHeader: J } = l(a[c]), B = [["class", `${f}-code`]];
      k && B.push(["open", ""]);
      const ne = {
        attrs: se(a[c], B)
      };
      a[c].info = a[c].info.replace(m, "");
      const W = r(a, c, $, p, g);
      return `
        <${R} ${g.renderAttrs(ne)}>
          <${J} class="${f}-code-head">
            <div class="${f}-code-flag"><span></span><span></span><span></span></div>
            <div class="${f}-code-action">
              <span class="${f}-code-lang">${e.utils.escapeHtml(a[c].info.trim())}</span>
              <span class="${f}-copy-button" data-tips="${h}"${y ? " data-is-icon=true" : ""}>${T}</span>
              ${t.extraTools instanceof Function ? t.extraTools({ lang: a[c].info.trim() }) : t.extraTools || ""}
              ${R === "details" ? b : ""}
            </div>
          </${J}>
          ${W}
        </${R}>
      `;
    }
    let w, F, x, L, _ = "", N = "", D = "";
    const { open: q, tagContainer: V, tagHeader: z } = l(a[c]), K = [["class", `${f}-code`]];
    q && K.push(["open", ""]);
    const Z = {
      attrs: se(a[c], K)
    };
    for (let k = c; k < a.length && (w = a[k], [F, x] = u(w), F === S); k++) {
      w.info = w.info.replace(o, "").replace(m, ""), w.hidden = !0;
      const R = `${f}-codetab-${t.editorId}-${c}-${k - c}`;
      L = k - c > 0 ? "" : "checked", _ += `
        <li>
          <input
            type="radio"
            id="label-${f}-codetab-label-1-${t.editorId}-${c}-${k - c}"
            name="${f}-codetab-label-${t.editorId}-${c}"
            class="${R}"
            ${L}
          >
          <label
            for="label-${f}-codetab-label-1-${t.editorId}-${c}-${k - c}"
            onclick="this.getRootNode().querySelectorAll('.${R}').forEach(e => e.click())"
          >
            ${e.utils.escapeHtml(x || s(w))}
          </label>
        </li>`, N += `
        <div role="tabpanel">
          <input
            type="radio"
            name="${f}-codetab-pre-${t.editorId}-${c}"
            class="${R}"
            ${L}
            role="presentation">
          ${r(a, k, $, p, g)}
        </div>`, D += `
        <input
          type="radio"
          name="${f}-codetab-lang-${t.editorId}-${c}"
          class="${R}"
          ${L}
          role="presentation">
        <span class=${f}-code-lang role="note">${e.utils.escapeHtml(s(w))}</span>`;
    }
    return `
      <${V} ${g.renderAttrs(Z)}>
        <${z} class="${f}-code-head">
          <div class="${f}-code-flag">
            <ul class="${f}-codetab-label" role="tablist">${_}</ul>
          </div>
          <div class="${f}-code-action">
            <span class="${f}-codetab-lang">${D}</span>
            <span class="${f}-copy-button" data-tips="${h}"${y ? " data-is-icon=true" : ""}>${T}</span>
            ${t.extraTools instanceof Function ? t.extraTools({ lang: a[c].info.trim() }) : t.extraTools || ""}
            ${V === "details" ? b : ""}
          </div>
        </${z}>
        ${N}
      </${V}>
    `;
  };
  e.renderer.rules.fence = i, e.renderer.rules.code_block = i;
}, ct = (e, t) => {
  const r = e.renderer.rules.fence.bind(e.renderer.rules);
  e.renderer.rules.fence = (n, o, m, d, u) => {
    const s = n[o], l = s.content.trim();
    if (s.info === "echarts") {
      if (s.attrSet("class", `${f}-echarts`), s.attrSet("data-echarts-theme", t.themeRef.value), s.map && s.level === 0) {
        const i = s.map[1] - 1, c = !!d.srcLines[i]?.trim()?.startsWith("```");
        s.attrSet("data-closed", `${c}`), s.attrSet("data-line", String(s.map[0]));
      }
      return `<div ${u.renderAttrs(s)} style="width: 100%; aspect-ratio: 4 / 3;">${e.utils.escapeHtml(l)}</div>`;
    }
    return r(n, o, m, d, u);
  };
}, dt = (e, t) => {
  e.renderer.rules.heading_open = (r, n) => {
    const o = r[n], m = r[n + 1].children?.reduce((u, s) => u + (["text", "code_inline", "math_inline"].includes(s.type) && s.content || ""), "") || "", d = o.markup.length;
    return t.headsRef.value.push({
      text: m,
      level: d,
      line: o.map[0],
      currentToken: o,
      nextToken: r[n + 1]
    }), o.map && o.level === 0 && o.attrSet(
      "id",
      t.mdHeadingId({
        text: m,
        level: d,
        index: t.headsRef.value.length,
        currentToken: o,
        nextToken: r[n + 1]
      })
    ), e.renderer.renderToken(r, n, t);
  }, e.renderer.rules.heading_close = (r, n, o, m, d) => d.renderToken(r, n, o);
}, be = {
  block: [
    { open: "$$", close: "$$" },
    { open: "\\[", close: "\\]" }
  ],
  inline: [
    { open: "$$", close: "$$" },
    { open: "$", close: "$" },
    { open: "\\[", close: "\\]" },
    { open: "\\(", close: "\\)" }
  ]
}, ut = (e) => (t, r) => {
  const n = e.delimiters;
  for (const o of n) {
    if (!t.src.startsWith(o.open, t.pos))
      continue;
    const m = t.pos + o.open.length;
    let d = m;
    for (; (d = t.src.indexOf(o.close, d)) !== -1; ) {
      let u = 0, s = d - 1;
      for (; s >= 0 && t.src[s] === "\\"; )
        u++, s--;
      if (u % 2 === 0)
        break;
      d += o.close.length;
    }
    if (d !== -1) {
      if (d - m === 0)
        return r || (t.pending += o.open + o.close), t.pos = d + o.close.length, !0;
      if (!r) {
        const u = t.push("math_inline", "math", 0);
        u.markup = o.open, u.content = t.src.slice(m, d);
      }
      return t.pos = d + o.close.length, !0;
    }
  }
  return !1;
}, mt = (e) => (t, r, n, o) => {
  const m = e.delimiters, d = t.bMarks[r] + t.tShift[r], u = t.eMarks[r], s = (l, i, a) => {
    t.line = i;
    const c = t.push("math_block", "math", 0);
    return c.block = !0, c.content = l, c.map = [r, t.line], c.markup = a, !0;
  };
  for (const l of m) {
    const i = d;
    if (t.src.slice(i, i + l.open.length) !== l.open)
      continue;
    const a = i + l.open.length, c = t.src.slice(a, u).trim(), $ = c === "", p = c === l.close, g = c.endsWith(l.close);
    if (!$ && !p && !g)
      continue;
    if (o)
      return !0;
    if (p)
      return s("", r + 1, l.open);
    if (!$ && g) {
      const w = c.slice(0, -l.close.length);
      return s(w, r + 1, l.open);
    }
    let h = r + 1, T = !1, y = "";
    for (; h < n; h++) {
      const w = t.bMarks[h] + t.tShift[h], F = t.eMarks[h];
      if (w < F && t.tShift[h] < t.blkIndent)
        break;
      if (t.src.slice(w, F).trim().endsWith(l.close)) {
        const L = t.src.slice(0, F).lastIndexOf(l.close);
        y = t.src.slice(w, L), T = !0;
        break;
      }
    }
    if (!T)
      continue;
    const S = t.getLines(r + 1, h, t.tShift[r], !0) + (y.trim() ? y : "");
    return s(S, h + 1, l.open);
  }
  return !1;
}, ft = (e, { katexRef: t, inlineDelimiters: r, blockDelimiters: n }) => {
  const o = (u, s, l, i, a = !1) => {
    const c = {
      attrs: se(u, [["class", s]])
    }, $ = i.renderAttrs(c);
    if (!t.value)
      return `<${l} ${$}>${u.content}</${l}>`;
    const p = t.value.renderToString(
      u.content,
      P.katexConfig({
        throwOnError: !1,
        displayMode: a
      })
    );
    return `<${l} ${$} data-processed>${p}</${l}>`;
  }, m = (u, s, l, i, a) => o(u[s], `${f}-katex-inline`, "span", a), d = (u, s, l, i, a) => o(u[s], `${f}-katex-block`, "p", a, !0);
  e.inline.ruler.before(
    "escape",
    "math_inline",
    ut({
      delimiters: r || be.inline
    })
  ), e.block.ruler.after(
    "blockquote",
    "math_block",
    mt({
      delimiters: n || be.block
    }),
    {
      alt: ["paragraph", "reference", "blockquote", "list"]
    }
  ), e.renderer.rules.math_inline = m, e.renderer.rules.math_block = d;
}, ht = (e, t) => {
  const r = e.renderer.rules.fence.bind(e.renderer.rules);
  e.renderer.rules.fence = (n, o, m, d, u) => {
    const s = n[o], l = s.content.trim();
    if (s.info === "mermaid") {
      if (s.attrSet("class", `${f}-mermaid`), s.attrSet("data-mermaid-theme", t.themeRef.value), s.map && s.level === 0) {
        const a = s.map[1] - 1, $ = !!d.srcLines[a]?.trim()?.startsWith("```");
        s.attrSet("data-closed", `${$}`), s.attrSet("data-line", String(s.map[0]));
      }
      const i = Y.get(l);
      return i ? (s.attrSet("data-processed", ""), s.attrSet("data-content", l), `<p ${u.renderAttrs(s)}>${i}</p>`) : `<div ${u.renderAttrs(s)}>${e.utils.escapeHtml(l)}</div>`;
    }
    return r(n, o, m, d, u);
  };
}, $e = (e, t, r) => {
  const n = e.attrIndex(t), o = [t, r];
  n < 0 ? e.attrPush(o) : (e.attrs = e.attrs || [], e.attrs[n] = o);
}, pt = (e) => e.type === "inline", gt = (e) => e.type === "paragraph_open", vt = (e) => e.type === "list_item_open", yt = (e) => e.content.indexOf("[ ] ") === 0 || e.content.indexOf("[x] ") === 0 || e.content.indexOf("[X] ") === 0, bt = (e, t) => pt(e[t]) && gt(e[t - 1]) && vt(e[t - 2]) && yt(e[t]), $t = (e, t) => {
  const r = e[t].level - 1;
  for (let n = t - 1; n >= 0; n--)
    if (e[n].level === r)
      return n;
  return -1;
}, wt = (e) => {
  const t = new e("html_inline", "", 0);
  return t.content = "<label>", t;
}, Ct = (e) => {
  const t = new e("html_inline", "", 0);
  return t.content = "</label>", t;
}, Et = (e, t, r) => {
  const n = new r("html_inline", "", 0);
  return n.content = '<label class="task-list-item-label" for="' + t + '">' + e + "</label>", n.attrs = [["for", t]], n;
}, kt = (e, t, r) => {
  const n = new t("html_inline", "", 0), o = r.enabled ? " " : ' disabled="" ';
  return e.content.indexOf("[ ] ") === 0 ? n.content = '<input class="task-list-item-checkbox"' + o + 'type="checkbox">' : (e.content.indexOf("[x] ") === 0 || e.content.indexOf("[X] ") === 0) && (n.content = '<input class="task-list-item-checkbox" checked=""' + o + 'type="checkbox">'), n;
}, Tt = (e, t, r) => {
  if (e.children = e.children || [], e.children.unshift(kt(e, t, r)), e.children[1].content = e.children[1].content.slice(3), e.content = e.content.slice(3), r.label)
    if (r.labelAfter) {
      e.children.pop();
      const n = "task-item-" + Math.ceil(Math.random() * (1e4 * 1e3) - 1e3);
      e.children[0].content = e.children[0].content.slice(0, -1) + ' id="' + n + '">', e.children.push(Et(e.content, n, t));
    } else
      e.children.unshift(wt(t)), e.children.push(Ct(t));
}, xt = (e, t = {}) => {
  e.core.ruler.after("inline", "github-task-lists", (r) => {
    const n = r.tokens;
    for (let o = 2; o < n.length; o++)
      bt(n, o) && (Tt(n[o], r.Token, t), $e(
        n[o - 2],
        "class",
        "task-list-item" + (t.enabled ? " enabled" : " ")
      ), $e(n[$t(n, o - 2)], "class", "contains-task-list"));
  });
}, It = (e) => {
  e.core.ruler.push("init-line-number", (t) => (t.tokens.forEach((r) => {
    r.map && (r.attrs || (r.attrs = []), r.attrs.push(["data-line", r.map[0].toString()]));
  }), !0));
}, St = (e, t) => {
  const { editorConfig: r, markdownItConfig: n, markdownItPlugins: o, editorExtensions: m } = P, d = C("editorId"), u = C("language"), s = C(
    "usedLanguageText"
  ), l = C("showCodeRowNumber"), i = C("theme"), a = C("customIcon"), c = C("rootRef"), $ = C("setting"), p = G([]), g = ot(e), h = at(e), { reRenderRef: T, replaceMermaid: y } = lt(e), { reRenderEcharts: b, replaceEcharts: S } = nt(e), w = Ke({
    html: !0,
    breaks: !0,
    linkify: !0
  });
  n(w, {
    editorId: d
  });
  const F = [
    {
      type: "image",
      plugin: Ze,
      options: { figcaption: !0, classes: "md-zoom" }
    },
    {
      type: "admonition",
      plugin: it,
      options: {}
    },
    {
      type: "taskList",
      plugin: xt,
      options: {}
    },
    {
      type: "heading",
      plugin: dt,
      options: { mdHeadingId: e.mdHeadingId, headsRef: p }
    },
    {
      type: "code",
      plugin: st,
      options: {
        editorId: d,
        usedLanguageTextRef: s,
        // showCodeRowNumber,
        codeFoldable: e.codeFoldable,
        autoFoldThreshold: e.autoFoldThreshold,
        customIconRef: a
      }
    },
    {
      type: "sub",
      plugin: Je,
      options: {}
    },
    {
      type: "sup",
      plugin: Xe,
      options: {}
    }
  ];
  e.noKatex || F.push({
    type: "katex",
    plugin: ft,
    options: { katexRef: h }
  }), e.noMermaid || F.push({
    type: "mermaid",
    plugin: ht,
    options: { themeRef: i }
  }), e.noEcharts || F.push({
    type: "echarts",
    plugin: ct,
    options: { themeRef: i }
  }), o(F, {
    editorId: d
  }).forEach((k) => {
    w.use(k.plugin, k.options);
  });
  const x = w.options.highlight;
  w.set({
    highlight: (k, R, J) => {
      if (x) {
        const W = x(k, R, J);
        if (W)
          return W;
      }
      let B;
      !e.noHighlight && g.value ? g.value.getLanguage(R) ? B = g.value.highlight(k, {
        language: R,
        ignoreIllegals: !0
      }).value : B = g.value.highlightAuto(k).value : B = w.utils.escapeHtml(k);
      const ne = l ? Ye(
        B.replace(/^\n+|\n+$/g, ""),
        k.replace(/^\n+|\n+$/g, "")
      ) : `<span class="${f}-code-block">${B.replace(/^\n+|\n+$/g, "")}</span>`;
      return `<pre><code class="language-${R}" language=${R}>${ne}</code></pre>`;
    }
  }), It(w);
  const L = G(`_article-key_${le()}`), _ = G(
    e.sanitize(
      w.render(e.modelValue, {
        srcLines: e.modelValue.split(`
`)
      })
    )
  );
  let N = () => {
  }, D = () => {
  };
  const q = () => {
    const k = c.value?.querySelectorAll(
      `#${d} p.${f}-mermaid:not([data-closed=false])`
    );
    D(), D = ze(k, {
      customIcon: a.value
    }), m.mermaid?.enableZoom && (N(), N = We(k, {
      customIcon: a.value
    }));
  }, V = () => {
    v.emit(d, X, _.value), e.onHtmlChanged(_.value), e.onGetCatalog(p.value), v.emit(d, oe, p.value), U(() => {
      y().then(q), S();
    });
  }, z = () => {
    p.value = [], _.value = e.sanitize(
      w.render(e.modelValue, {
        srcLines: e.modelValue.split(`
`)
      })
    );
  }, K = A(() => (e.noKatex || !!h.value) && (e.noHighlight || !!g.value));
  let Z = -1;
  return E([Ae(e, "modelValue"), K, T, u], () => {
    Z = window.setTimeout(
      () => {
        z();
      },
      t ? 0 : r.renderDelay
    );
  }), E(
    () => $.value.preview,
    () => {
      $.value.preview && U(() => {
        y().then(q), S(), v.emit(d, oe, p.value);
      });
    }
  ), E([_, b], () => {
    V();
  }), H(V), H(() => {
    v.on(d, {
      name: De,
      callback() {
        v.emit(d, oe, p.value);
      }
    }), v.on(d, {
      name: de,
      callback: () => {
        L.value = `_article-key_${le()}`, z();
      }
    });
  }), te(() => {
    N(), D(), clearTimeout(Z);
  }), { html: _, key: L };
}, Ft = (e, t) => {
  const r = C("editorId"), n = C("setting"), { noImgZoomIn: o } = e, m = Le(() => {
    const d = document.querySelectorAll(
      `#${r}-preview img:not(.not-zoom):not(.medium-zoom-image)`
    );
    d.length !== 0 && Qe(d, {
      background: "#00000073"
    });
  });
  H(async () => {
    !o && n.value.preview && await m();
  }), E([t, () => n.value.preview], async () => {
    !o && n.value.preview && await m();
  });
}, we = {
  checked: {
    regexp: /- \[x\]/,
    value: "- [ ]"
  },
  unChecked: {
    regexp: /- \[\s\]/,
    value: "- [x]"
  }
}, At = (e, t) => {
  const r = C("editorId"), n = C("rootRef");
  let o = () => {
  };
  const m = () => {
    if (!n.value)
      return !1;
    const d = n.value.querySelectorAll(".task-list-item.enabled"), u = (s) => {
      s.preventDefault();
      const l = s.target.checked ? "unChecked" : "checked", i = s.target.parentElement?.dataset.line;
      if (!i)
        return;
      const a = Number(i), c = e.modelValue.split(`
`), $ = c[Number(a)].replace(
        we[l].regexp,
        we[l].value
      );
      e.previewOnly ? (c[Number(a)] = $, e.onChange(c.join(`
`))) : v.emit(r, Ve, a + 1, $);
    };
    d.forEach((s) => {
      s.addEventListener("click", u);
    }), o = () => {
      d.forEach((s) => {
        s.removeEventListener("click", u);
      });
    };
  };
  te(() => {
    o();
  }), E(
    [t],
    () => {
      o(), U(m);
    },
    {
      immediate: !0
    }
  );
}, Rt = (e, t, r) => {
  const n = C("setting"), o = () => {
    U(() => {
      e.onRemount?.();
    });
  }, m = (d) => {
    d && o();
  };
  E([t, r], o), E(() => n.value.preview, m), E(() => n.value.htmlPreview, m), H(o);
}, xe = {
  modelValue: {
    type: String,
    default: ""
  },
  onChange: {
    type: Function,
    default: () => {
    }
  },
  onHtmlChanged: {
    type: Function,
    default: () => {
    }
  },
  onGetCatalog: {
    type: Function,
    default: () => {
    }
  },
  mdHeadingId: {
    type: Function,
    default: () => ""
  },
  noMermaid: {
    type: Boolean,
    default: !1
  },
  sanitize: {
    type: Function,
    default: (e) => e
  },
  // 不使用该函数功能
  noKatex: {
    type: Boolean,
    default: !1
  },
  formatCopiedText: {
    type: Function,
    default: (e) => e
  },
  noHighlight: {
    type: Boolean,
    default: !1
  },
  previewOnly: {
    type: Boolean,
    default: !1
  },
  noImgZoomIn: {
    type: Boolean
  },
  sanitizeMermaid: {
    type: Function
  },
  codeFoldable: {
    type: Boolean
  },
  autoFoldThreshold: {
    type: Number
  },
  onRemount: {
    type: Function
  },
  noEcharts: {
    type: Boolean
  },
  previewComponent: {
    type: [Object, Function],
    default: void 0
  }
}, nn = {
  ...xe,
  updateModelValue: {
    type: Function,
    default: () => {
    }
  },
  placeholder: {
    type: String,
    default: ""
  },
  scrollAuto: {
    type: Boolean
  },
  autofocus: {
    type: Boolean
  },
  readonly: {
    type: Boolean
  },
  maxlength: {
    type: Number
  },
  autoDetectCode: {
    type: Boolean
  },
  /**
   * 输入框失去焦点时触发事件
   */
  onBlur: {
    type: Function,
    default: () => {
    }
  },
  /**
   * 输入框获得焦点时触发事件
   */
  onFocus: {
    type: Function,
    default: () => {
    }
  },
  completions: {
    type: Array
  },
  onInput: {
    type: Function
  },
  onDrop: {
    type: Function,
    default: () => {
    }
  },
  inputBoxWidth: {
    type: String
  },
  oninputBoxWidthChange: {
    type: Function
  },
  transformImgUrl: {
    type: Function,
    default: (e) => e
  },
  catalogLayout: {
    type: String
  },
  catalogMaxDepth: {
    type: Number
  }
}, Ce = (e) => {
  const r = new DOMParser().parseFromString(e, "text/html");
  return Array.from(r.body.childNodes);
}, Ht = (e, t) => e.nodeType !== t.nodeType ? !1 : e.nodeType === Node.TEXT_NODE || e.nodeType === Node.COMMENT_NODE ? e.textContent === t.textContent : e.nodeType === Node.ELEMENT_NODE ? e.outerHTML === t.outerHTML : e.isEqualNode ? e.isEqualNode(t) : !1, Mt = /* @__PURE__ */ ce({
  name: "UpdateOnDemand",
  props: {
    id: {
      type: String,
      required: !0
    },
    class: {
      type: [String, Array, Object],
      required: !0
    },
    html: {
      type: String,
      required: !0
    }
  },
  setup(e) {
    const t = G(), r = e.html, n = (o, m) => {
      if (!t.value) return;
      const d = t.value, u = Array.from(d.childNodes), s = Math.min(o.length, m.length);
      let l = -1;
      for (let a = 0; a < s; a++)
        if (!Ht(o[a], m[a])) {
          l = a;
          break;
        }
      if (l === -1)
        if (m.length > o.length)
          l = o.length;
        else if (o.length > m.length)
          l = m.length;
        else
          return;
      const i = Math.min(l, u.length);
      for (let a = u.length - 1; a >= i; a--)
        u[a].remove();
      for (let a = l; a < o.length; a++)
        d.appendChild(o[a].cloneNode(!0));
    };
    return E(() => e.html, (o, m) => {
      const d = Ce(o), u = Ce(m || "");
      n(d, u);
    }), () => j("div", {
      id: e.id,
      class: e.class,
      innerHTML: r,
      ref: t
    }, null);
  }
}), Pt = /* @__PURE__ */ ce({
  name: "ContentPreview",
  props: xe,
  setup(e) {
    const t = C("editorId"), r = C("setting"), n = C("previewTheme"), o = C("showCodeRowNumber"), {
      html: m,
      key: d
    } = St(e, e.previewOnly);
    tt(e, m, d), Ft(e, m), At(e, m), Rt(e, m, d);
    const u = A(() => [`${f}-preview`, `${n?.value}-theme`, o && `${f}-scrn`].filter(Boolean)), s = () => {
      const l = `${t}-preview`;
      return e.previewComponent ? He(e.previewComponent, {
        key: d.value,
        html: m.value,
        id: l,
        class: u.value
      }) : j(Mt, {
        key: d.value,
        html: m.value,
        id: l,
        class: u.value
      }, null);
    };
    return () => j(Re, null, [r.value.preview && (e.previewOnly ? s() : j("div", {
      id: `${t}-preview-wrapper`,
      class: `${f}-preview-wrapper`,
      key: "content-preview-wrapper"
    }, [s()])), r.value.htmlPreview && j("div", {
      id: `${t}-html-wrapper`,
      class: `${f}-preview-wrapper`,
      key: "html-preview-wrapper"
    }, [j("div", {
      class: `${f}-html`
    }, [m.value])])]);
  }
}), Lt = ({ text: e }) => e, Ie = {
  /**
   * markdown content.
   *
   * @default ''
   */
  modelValue: {
    type: String,
    default: ""
  },
  /**
   * input回调事件
   */
  onChange: {
    type: Function,
    default: void 0
  },
  /**
   * 主题，支持light和dark
   *
   * @default 'light'
   */
  theme: {
    type: String,
    default: "light"
  },
  /**
   * 外层类名
   *
   * @default ''
   */
  class: {
    type: String,
    default: ""
  },
  /**
   * 预设语言名称
   *
   * @default 'zh-CN'
   */
  language: {
    type: String,
    default: "zh-CN"
  },
  /**
   * html变化事件
   */
  onHtmlChanged: {
    type: Function,
    default: void 0
  },
  /**
   * 获取目录结构
   */
  onGetCatalog: {
    type: Function,
    default: void 0
  },
  /**
   * 编辑器唯一标识
   *
   * @default 'md-editor-v3'
   * @deprecated 5.x版本开始使用 id 替换
   */
  editorId: {
    type: String,
    default: void 0
  },
  /**
   * 5.x版本开始 editorId 的替换
   *
   * @default 'md-editor-v3'
   */
  id: {
    type: String,
    default: void 0
  },
  /**
   * 预览中代码是否显示行号
   *
   * @default true
   */
  showCodeRowNumber: {
    type: Boolean,
    default: !0
  },
  /**
   * 预览内容样式
   *
   * @default 'default'
   */
  previewTheme: {
    type: String,
    default: "default"
  },
  /**
   * 编辑器样式
   */
  style: {
    type: Object,
    default: () => ({})
  },
  /**
   * 标题的id生成方式
   *
   * @default (text: string) => text
   */
  mdHeadingId: {
    type: Function,
    default: Lt
  },
  /**
   *
   * 不能保证文本正确的情况，在marked编译md文本后通过该方法处理
   * 推荐DOMPurify、sanitize-html
   *
   * @default (text: string) => text
   */
  sanitize: {
    type: Function,
    default: (e) => e
  },
  /**
   * 不使用该mermaid
   *
   * @default false
   */
  noMermaid: {
    type: Boolean,
    default: !1
  },
  /**
   * 不使用katex
   *
   * @default false
   */
  noKatex: {
    type: Boolean,
    default: !1
  },
  /**
   * 代码主题
   *
   * @default 'atom'
   */
  codeTheme: {
    type: String,
    default: "atom"
  },
  /**
   * 复制代码格式化方法
   *
   * @default (text) => text
   */
  formatCopiedText: {
    type: Function,
    default: (e) => e
  },
  /**
   * 某些预览主题的代码模块背景是暗色系
   * 将这个属性设置为true，会自动在该主题下的light模式下使用暗色系的代码风格
   *
   * @default true
   */
  codeStyleReverse: {
    type: Boolean,
    default: !0
  },
  /**
   * 需要自动调整的预览主题
   *
   * @default ['default', 'mk-cute']
   */
  codeStyleReverseList: {
    type: Array,
    default: ["default", "mk-cute"]
  },
  noHighlight: {
    type: Boolean,
    default: !1
  },
  /**
   * 是否关闭编辑器默认的放大缩小功能
   */
  noImgZoomIn: {
    type: Boolean,
    default: !1
  },
  /**
   * 自定义的图标
   */
  customIcon: {
    type: Object,
    default: {}
  },
  sanitizeMermaid: {
    type: Function,
    default: (e) => Promise.resolve(e)
  },
  /**
   * 是否开启折叠代码功能
   * 不开启会使用div标签替代details标签
   *
   * @default true
   */
  codeFoldable: {
    type: Boolean,
    default: !0
  },
  /**
   * 触发自动折叠代码的行数阈值
   *
   * @default 30
   */
  autoFoldThreshold: {
    type: Number,
    default: 30
  },
  /**
   * 内容重新挂载事件
   *
   * 相比起onHtmlChanged，onRemount会在重新挂载后触发
   */
  onRemount: {
    type: Function,
    default: void 0
  },
  /**
   * 不使用 echarts
   */
  noEcharts: {
    type: Boolean,
    default: !1
  },
  previewComponent: {
    type: [Object, Function],
    default: void 0
  }
}, on = {
  ...Ie,
  /**
   * input回调事件
   */
  onSave: {
    type: Function,
    default: void 0
  },
  /**
   * 上传图片事件
   */
  onUploadImg: {
    type: Function,
    default: void 0
  },
  /**
   * 是否页面内全屏
   *
   * @default false
   */
  pageFullscreen: {
    type: Boolean,
    default: !1
  },
  /**
   * 是否展开预览
   *
   * @default true
   */
  preview: {
    type: Boolean,
    default: !0
  },
  /**
   * 是否展开html预览
   *
   * @default false
   */
  htmlPreview: {
    type: Boolean,
    default: !1
  },
  /**
   * 仅预览模式，不显示toolbar和编辑框
   *
   * @4.0.0开始移除该设置，使用组件MdPreview替换
   *
   * @default false
   */
  // previewOnly: {
  //   type: Boolean as PropType<boolean>,
  //   default: false
  // },
  /**
   * 工具栏选择显示
   *
   * @default allToolbar
   */
  toolbars: {
    type: Array,
    default: _e
  },
  /**
   * 浮动工具栏
   *
   * @version 6.0.0
   * @default []
   */
  floatingToolbars: {
    type: Array,
    default: []
  },
  /**
   * 工具栏选择不显示
   *
   * @default []
   */
  toolbarsExclude: {
    type: Array,
    default: []
  },
  /**
   * 格式化md
   *
   * @default true
   */
  noPrettier: {
    type: Boolean,
    default: !1
  },
  /**
   * 一个tab等于空格数
   *
   * @default 2
   */
  tabWidth: {
    type: Number,
    default: 2
  },
  /**
   * 表格预设格子数
   *
   * 也可以是[6, 4, 10, 8]
   *
   * @default [6, 4]
   */
  tableShape: {
    type: Array,
    default: [6, 4]
  },
  /**
   * 空提示
   *
   * @default ''
   */
  placeholder: {
    type: String,
    default: ""
  },
  /**
   * 自定义的工具栏列表
   */
  defToolbars: {
    type: [String, Object],
    default: void 0
  },
  /**
   * 内部错误捕获
   */
  onError: {
    type: Function,
    default: void 0
  },
  /**
   * 页脚列表显示顺序
   */
  footers: {
    type: Array,
    default: Oe
  },
  /**
   * 是否默认激活输入框和预览框同步滚动
   *
   * @default true
   */
  scrollAuto: {
    type: Boolean,
    default: !0
  },
  /**
   * 自定义的也叫工具组件列表
   */
  defFooters: {
    type: [String, Object],
    default: void 0
  },
  /**
   * 是否禁用上传图片
   *
   * @default false
   */
  noUploadImg: {
    type: Boolean,
    default: !1
  },
  /**
   * 文本区域自动获得焦点
   *
   * @default false
   */
  autoFocus: {
    type: Boolean,
    default: !1
  },
  /**
   * 禁用文本区域
   *
   * @default false
   */
  disabled: {
    type: Boolean,
    default: !1
  },
  /**
   * 文本区域为只读
   *
   * @default false
   */
  readOnly: {
    type: Boolean,
    default: !1
  },
  /**
   * 文本区域允许的最大字符数
   */
  maxLength: {
    type: Number,
    default: void 0
  },
  /**
   * 是否启用自动识别粘贴代码类别
   * 目前支持 vscode 复制的代码识别
   *
   * @default false
   */
  autoDetectCode: {
    type: Boolean,
    default: !1
  },
  /**
   * 输入框失去焦点时触发事件
   */
  onBlur: {
    type: Function,
    default: void 0
  },
  /**
   * 输入框获得焦点时触发事件
   */
  onFocus: {
    type: Function,
    default: void 0
  },
  /**
   * @codemirror/autocomplete匹配关键词的方法列表
   *
   * 它会被像下面这样嵌入编辑器
   *
   * import { autocompletion } from '@codemirror/autocomplete';
   * autocompletion({
   *   override: [...completions]
   * })
   */
  completions: {
    type: Array,
    default: void 0
  },
  /**
   * 是否在工具栏下面显示对应的文字名称
   *
   * @default false
   */
  showToolbarName: {
    type: Boolean,
    default: !1
  },
  /**
   * 字符输入事件
   */
  onInput: {
    type: Function,
    default: void 0
  },
  onDrop: {
    type: Function,
    default: void 0
  },
  /**
   * 输入框的默认宽度
   *
   * @example '100px'/'50%'
   * @default '50%
   */
  inputBoxWidth: {
    type: String,
    default: "50%"
  },
  /**
   * 输入框宽度变化事件
   */
  oninputBoxWidthChange: {
    type: Function,
    default: void 0
  },
  /**
   * 替换粘贴的图片链接
   *
   * @param t 图片链接
   * @returns
   */
  transformImgUrl: {
    type: Function,
    default: (e) => e
  },
  /**
   * 内置的目录显示的状态
   *
   * 'fixed': 悬浮在内容上方
   * 'flat': 展示在右侧
   *
   * \>=5.3.0
   *
   * @default 'fixed'
   */
  catalogLayout: {
    type: String,
    default: "fixed"
  },
  /**
   * 控制最大显示的目录层级
   */
  catalogMaxDepth: {
    type: Number,
    default: void 0
  }
}, Se = [
  "onHtmlChanged",
  "onGetCatalog",
  "onChange",
  "onRemount",
  "update:modelValue"
], an = [
  ...Se,
  "onSave",
  "onUploadImg",
  "onError",
  "onBlur",
  "onFocus",
  "onInput",
  "onDrop",
  "oninputBoxWidthChange"
], Ot = (e, t, r) => {
  const { editorId: n } = r, o = {
    rerender() {
      v.emit(n, de);
    }
  };
  t.expose(o);
}, ae = /* @__PURE__ */ ce({
  name: "MdPreview",
  props: Ie,
  emits: Se,
  setup(e, t) {
    const {
      noKatex: r,
      noMermaid: n,
      noHighlight: o
    } = e, m = G(), d = et(e);
    Te(e, {
      rootRef: m,
      editorId: d
    }), Ot(e, t, {
      editorId: d
    }), te(() => {
      v.clear(d);
    });
    const u = (a) => {
      e.onChange?.(a), t.emit("onChange", a), t.emit("update:modelValue", a);
    }, s = (a) => {
      e.onHtmlChanged?.(a), t.emit("onHtmlChanged", a);
    }, l = (a) => {
      e.onGetCatalog?.(a), t.emit("onGetCatalog", a);
    }, i = () => {
      e.onRemount?.(), t.emit("onRemount");
    };
    return () => j("div", {
      id: d,
      class: [f, e.class, e.theme === "dark" && `${f}-dark`, `${f}-previewOnly`],
      style: e.style,
      ref: m
    }, [j(Pt, {
      modelValue: e.modelValue,
      onChange: u,
      onHtmlChanged: s,
      onGetCatalog: l,
      mdHeadingId: e.mdHeadingId,
      noMermaid: n,
      sanitize: e.sanitize,
      noKatex: r,
      formatCopiedText: e.formatCopiedText,
      noHighlight: o,
      noImgZoomIn: e.noImgZoomIn,
      previewOnly: !0,
      sanitizeMermaid: e.sanitizeMermaid,
      codeFoldable: e.codeFoldable,
      autoFoldThreshold: e.autoFoldThreshold,
      onRemount: i,
      noEcharts: e.noEcharts,
      previewComponent: e.previewComponent
    }, null)]);
  }
});
ae.install = (e) => (e.component(ae.name, ae), e);
export {
  M as C,
  ae as M,
  Pt as a,
  on as b,
  nn as c,
  Qt as d,
  an as e,
  en as f,
  Zt as g,
  Xt as h,
  Yt as i,
  tn as j,
  Jt as k,
  et as u
};
